File: backends/kotlin/zappy-adapters/src/main/kotlin/nl/zappymart/adapters/seed/SeedLoader.kt
package nl.zappymart.adapters.seed

import jakarta.persistence.EntityManager
import nl.zappymart.adapters.persistence.CachedProductRepository
import nl.zappymart.adapters.persistence.entities.CategoryEntity
import nl.zappymart.adapters.persistence.entities.CustomerEntity
import nl.zappymart.adapters.persistence.entities.ProductEntity
import nl.zappymart.adapters.persistence.entities.PromotionEntity
import nl.zappymart.adapters.persistence.entities.WishlistEntryEntity
import nl.zappymart.application.ports.PasswordHasher
import nl.zappymart.domain.shared.PromotionCode
import org.springframework.core.io.ClassPathResource
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import tools.jackson.core.type.TypeReference
import tools.jackson.databind.ObjectMapper

@Component
class SeedLoader(
    private val entityManager: EntityManager,
    private val objectMapper: ObjectMapper,
    private val passwords: PasswordHasher,
    private val cachedProducts: CachedProductRepository,
) {

    @Transactional
    fun load(): Int {
        emptyTheStore()
        val categories = readSeedFile<SeedCategory>("categories.json")
        val products = readSeedFile<SeedProduct>("products.json")
        val promotionCodes = readSeedFile<SeedPromotionCode>("promotion-codes.json")
        val customers = readSeedFile<SeedCustomer>("customers.json")

        val categoryBySlug = categories.mapIndexed { position, category ->
            val entity = CategoryEntity(category.id, category.name, category.slug, position)
            entityManager.persist(entity)
            category.slug to entity
        }.toMap()

        products.forEachIndexed { position, product ->
            entityManager.persist(
                ProductEntity(
                    id = product.id,
                    name = product.name,
                    slug = product.slug,
                    description = product.description,
                    priceAmount = product.price.amount,
                    currency = product.price.currency,
                    category = requireNotNull(categoryBySlug[product.categorySlug]) {
                        "The seed puts ${product.id} in the unknown category ${product.categorySlug}"
                    },
                    stock = product.stock,
                    imageUrl = product.imageUrl,
                    cataloguePosition = position,
                ),
            )
        }

        promotionCodes.forEach { code ->
            entityManager.persist(
                PromotionEntity(
                    code = PromotionCode.of(code.code).value,
                    kind = code.kind,
                    percentage = code.percentage,
                    amount = code.amount?.amount,
                    minimumSubtotal = code.minimumSubtotal?.amount,
                    currency = code.amount?.currency ?: code.minimumSubtotal?.currency ?: "EUR",
                    validFrom = code.validFrom,
                    validUntil = code.validUntil,
                    usageLimit = code.usageLimit,
                    timesUsed = code.timesUsed,
                ),
            )
        }

        customers.forEach { customer ->
            entityManager.persist(
                CustomerEntity(
                    id = customer.id,
                    email = customer.email.trim().lowercase(),
                    name = customer.name,
                    passwordHash = passwords.hash(customer.password).value,
                    createdAt = customer.createdAt,
                ),
            )
            customer.wishlist.forEachIndexed { position, productId ->
                entityManager.persist(WishlistEntryEntity(ownerId = customer.id, productId = productId, position = position))
            }
        }

        entityManager.flush()
        entityManager.clear()
        cachedProducts.forget()
        return products.size
    }

    private fun emptyTheStore() {
        TABLES_IN_DELETION_ORDER.forEach { entityName ->
            entityManager.createQuery("delete from $entityName").executeUpdate()
        }
        entityManager.flush()
        entityManager.clear()
        cachedProducts.forget()
    }

    private inline fun <reified Item> readSeedFile(name: String): List<Item> =
        ClassPathResource("$SEED_LOCATION/$name").inputStream.use { stream ->
            objectMapper.readValue(stream, object : TypeReference<List<Item>>() {})
        }

    private companion object {
        const val SEED_LOCATION = "contract-seed"

        val TABLES_IN_DELETION_ORDER = listOf(
            "OrderLineEntity",
            "OrderEntity",
            "CartLineEntity",
            "CartEntity",
            "RefreshTokenEntity",
            "SessionEntity",
            "WishlistEntryEntity",
            "PromotionEntity",
            "ProductEntity",
            "CategoryEntity",
            "CustomerEntity",
        )
    }
}

File: backends/dotnet/src/Zappy.Adapters.Persistence/Seed/SeedLoader.cs
using System.Text.Json;
using Microsoft.EntityFrameworkCore;
using Zappy.Application;
using Zappy.Domain;

namespace Zappy.Adapters.Persistence;

public sealed class SeedLoader(
    ZappyDbContext database,
    IPasswordHasher passwordHasher,
    ProductCatalogueVersion catalogueVersion,
    SeedSettings settings) : ISeedLoader
{
    private static readonly JsonSerializerOptions ReadingOptions = new(JsonSerializerDefaults.Web);

    public async Task<int> LoadFreshSeed(CancellationToken cancellationToken)
    {
        await EmptyTheStore(cancellationToken);

        var folder = SeedDirectory.Find(settings.Directory);
        var categories = await Read<SeedCategory>(Path.Combine(folder, "categories.json"), cancellationToken);
        var products = await Read<SeedProduct>(Path.Combine(folder, "products.json"), cancellationToken);
        var promotionCodes = await Read<SeedPromotionCode>(Path.Combine(folder, "promotion-codes.json"), cancellationToken);
        var customers = await Read<SeedCustomer>(Path.Combine(folder, "customers.json"), cancellationToken);

        for (var position = 0; position < categories.Count; position += 1)
        {
            var category = categories[position];
            database.Categories.Add(new Category(category.Id, category.Name, category.Slug, position + 1));
        }

        for (var position = 0; position < products.Count; position += 1)
        {
            var product = products[position];
            database.Products.Add(new Product(
                product.Id,
                product.Name,
                product.Slug,
                product.Description,
                new Money(product.Price.Amount, product.Price.Currency),
                product.CategorySlug,
                product.Stock,
                product.ImageUrl,
                position + 1));
        }

        foreach (var promotionCode in promotionCodes)
        {
            database.PromotionCodes.Add(new PromotionCode(
                promotionCode.Code,
                KindOf(promotionCode.Kind),
                promotionCode.Percentage,
                MoneyOf(promotionCode.Amount),
                MoneyOf(promotionCode.MinimumSubtotal),
                promotionCode.ValidFrom,
                promotionCode.ValidUntil,
                promotionCode.UsageLimit,
                promotionCode.TimesUsed));
        }

        foreach (var customer in customers)
        {
            var email = EmailAddress.Create(customer.Email)
                ?? throw new InvalidOperationException($"The seed customer {customer.Id} has an invalid email address.");

            database.Customers.Add(new Customer(
                customer.Id,
                email,
                customer.Name,
                passwordHasher.Hash(customer.Password),
                customer.CreatedAt));

            foreach (var productId in customer.Wishlist)
            {
                database.WishlistEntries.Add(new WishlistEntry(customer.Id, productId, customer.CreatedAt));
            }
        }

        await database.SaveChangesAsync(cancellationToken);
        catalogueVersion.Bump();
        return products.Count;
    }

    private async Task EmptyTheStore(CancellationToken cancellationToken)
    {
        await database.OrderLines.ExecuteDeleteAsync(cancellationToken);
        await database.Orders.ExecuteDeleteAsync(cancellationToken);
        await database.CartLines.ExecuteDeleteAsync(cancellationToken);
        await database.Carts.ExecuteDeleteAsync(cancellationToken);
        await database.WishlistEntries.ExecuteDeleteAsync(cancellationToken);
        await database.RefreshTokens.ExecuteDeleteAsync(cancellationToken);
        await database.Sessions.ExecuteDeleteAsync(cancellationToken);
        await database.Customers.ExecuteDeleteAsync(cancellationToken);
        await database.Products.ExecuteDeleteAsync(cancellationToken);
        await database.PromotionCodes.ExecuteDeleteAsync(cancellationToken);
        await database.Categories.ExecuteDeleteAsync(cancellationToken);
        database.ChangeTracker.Clear();
    }

    private static async Task<IReadOnlyList<TSeed>> Read<TSeed>(string path, CancellationToken cancellationToken)
    {
        await using var file = File.OpenRead(path);
        return await JsonSerializer.DeserializeAsync<List<TSeed>>(file, ReadingOptions, cancellationToken)
            ?? throw new InvalidOperationException($"The seed file {path} holds no list.");
    }

    private static Money? MoneyOf(SeedMoney? money) => money is null ? null : new Money(money.Amount, money.Currency);

    private static PromotionKind KindOf(string kind) => kind switch
    {
        "PERCENTAGE" => PromotionKind.Percentage,
        "FIXED_AMOUNT" => PromotionKind.FixedAmount,
        "FREE_SHIPPING" => PromotionKind.FreeShipping,
        _ => throw new InvalidOperationException($"The seed names the unknown promotion kind {kind}.")
    };
}
